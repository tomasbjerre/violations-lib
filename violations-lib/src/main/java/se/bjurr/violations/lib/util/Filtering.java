package se.bjurr.violations.lib.util;

import static se.bjurr.violations.lib.model.SEVERITY.ERROR;
import static se.bjurr.violations.lib.model.SEVERITY.INFO;
import static se.bjurr.violations.lib.model.SEVERITY.WARN;

import java.util.Set;
import java.util.TreeSet;
import se.bjurr.violations.lib.model.SEVERITY;
import se.bjurr.violations.lib.model.Violation;

public final class Filtering {
  private Filtering() {}

  public static Set<Violation> withAtLEastSeverity(
      final Set<Violation> unfiltered, final SEVERITY severity) {
    final Set<Violation> filtered = new TreeSet<>();
    for (final Violation candidate : unfiltered) {
      if (isSeverer(candidate.getSeverity(), severity)) {
        filtered.add(candidate);
      }
    }
    return filtered;
  }

  private static boolean isSeverer(final SEVERITY candiate, final SEVERITY atLeast) {
    return candiate == ERROR //
        || atLeast == INFO //
        || candiate == WARN && atLeast == WARN;
  }
}
