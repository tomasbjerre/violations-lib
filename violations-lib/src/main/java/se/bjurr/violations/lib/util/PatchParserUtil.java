package se.bjurr.violations.lib.util;

import static java.util.Optional.empty;
import static java.util.Optional.of;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import java.util.Map;
import java.util.Optional;
import java.util.TreeMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@SuppressFBWarnings("REDOS")
public class PatchParserUtil {

  private static final Pattern RANGE_PATTERN =
      Pattern.compile(
          "@@\\p{IsWhite_Space}-([0-9]+)(?:,[0-9]+)?\\p{IsWhite_Space}\\+([0-9]+)(?:,[0-9]+)?\\p{IsWhite_Space}@@.*");

  private final Map<Integer, Optional<Integer>> newLineToOldLineTable;
  private final Map<Integer, Optional<Integer>> newLineToLineInDiffTable;

  private final String patchString;

  public PatchParserUtil(final String patchString) {
    this.patchString = patchString;
    this.newLineToOldLineTable = new TreeMap<>();
    this.newLineToLineInDiffTable = new TreeMap<>();
    if (patchString == null) {
      return;
    }
    int currentLine = -1;
    int patchLocation = 1;
    int diffLocation = 0;
    for (final String line : patchString.split("\n")) {
      if (line.startsWith("@")) {
        final Matcher matcher = RANGE_PATTERN.matcher(line);
        if (!matcher.matches()) {
          throw new IllegalStateException(
              "Unable to parse patch line " + line + "\nFull patch: \n" + patchString);
        }
        currentLine = Integer.parseInt(matcher.group(2));
        patchLocation = Integer.parseInt(matcher.group(1));
      } else if (line.startsWith("+") && !line.startsWith("++")) {
        this.newLineToOldLineTable.put(currentLine, empty());
        currentLine++;
      } else if (line.startsWith(" ")) {
        this.newLineToOldLineTable.put(currentLine, of(patchLocation));
        currentLine++;
        patchLocation++;
      } else {
        patchLocation++;
      }
      diffLocation++;
      this.newLineToLineInDiffTable.put(currentLine, of(diffLocation));
    }
  }

  public boolean isLineInDiff(final Integer newLine) {
    return this.newLineToOldLineTable.containsKey(newLine);
  }

  public Optional<Integer> findOldLine(final Integer newLine) {
    if (this.newLineToOldLineTable.containsKey(newLine)) {
      return this.newLineToOldLineTable.get(newLine);
    }
    return empty();
  }

  public Optional<Integer> findLineInDiff(final int newLine) {
    if (this.newLineToLineInDiffTable.containsKey(newLine)) {
      return this.newLineToLineInDiffTable.get(newLine);
    }
    return empty();
  }

  Map<Integer, Optional<Integer>> getNewLineToOldLineTable() {
    return this.newLineToOldLineTable;
  }

  @Override
  public String toString() {
    final StringBuilder sb = new StringBuilder();
    sb.append("patch:\n");
    this.withLines(sb, this.patchString);
    sb.append("\n\n");
    sb.append("newLineToLineInDiffTable:\n");
    this.toString(sb, this.newLineToLineInDiffTable);
    sb.append("\n\nnewLineToOldLineTable:\n");
    this.toString(sb, this.newLineToOldLineTable);
    return sb.toString();
  }

  private void withLines(final StringBuilder sb, final String str) {
    final String[] lines = str.split("\n");
    for (int i = 0; i < lines.length; i++) {
      sb.append(i + "\t: " + lines[i] + "\n");
    }
  }

  private void toString(final StringBuilder sb, final Map<Integer, Optional<Integer>> map) {
    for (final Map.Entry<Integer, Optional<Integer>> e : map.entrySet()) {
      sb.append(e.getKey() + " : " + e.getValue().orElse(null) + "\n");
    }
  }
}
