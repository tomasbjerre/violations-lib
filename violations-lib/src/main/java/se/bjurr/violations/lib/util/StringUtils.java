package se.bjurr.violations.lib.util;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import java.util.Collections;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class StringUtils {
  private static final Map<String, Character> XML_ESCAPE_CHARACTER_MAP;
  private static final String[] XML_ESCAPE_CHARACTERS;
  private static final char XML_ESCAPE_START = '&';

  static {
    final Map<String, Character> temp = new ConcurrentHashMap<>();
    temp.put("&apos;", '\'');
    temp.put("&quot;", '\"');
    temp.put("&amp;", '&');
    temp.put("&lt;", '<');
    temp.put("&gt;", '>');
    XML_ESCAPE_CHARACTER_MAP = Collections.unmodifiableMap(temp);
    XML_ESCAPE_CHARACTERS = temp.keySet().toArray(new String[0]);
  }

  private StringUtils() {}

  public static String htmlDecode(final String input) {
    final Pattern p = Pattern.compile("&#(\\d+);");
    final Matcher m = p.matcher(input);
    final StringBuffer sb = new StringBuffer();
    while (m.find()) {
      final Integer found = Integer.valueOf(m.group(1));
      final String character = String.valueOf((char) found.intValue());
      m.appendReplacement(sb, character);
    }
    m.appendTail(sb);
    return sb.toString();
  }

  public static String xmlDecode(final String input) {
    final StringBuilder result = new StringBuilder(input.length());
    for (int i = 0; i < input.length(); ) {
      final char current = input.charAt(i);

      boolean isValidXmlEscape = false;
      if (current == XML_ESCAPE_START) {
        for (final String escapeCharacter : XML_ESCAPE_CHARACTERS) {
          if (input.startsWith(escapeCharacter, i)) {
            result.append(XML_ESCAPE_CHARACTER_MAP.get(escapeCharacter));
            i += escapeCharacter.length();
            isValidXmlEscape = true;
            break;
          }
        }
      }

      if (!isValidXmlEscape) {
        result.append(current);
        i++;
      }
    }
    return result.toString();
  }

  public static String escapeHTML(final String s) {
    final StringBuilder out = new StringBuilder(Math.max(16, s.length()));
    for (int i = 0; i < s.length(); i++) {
      final char c = s.charAt(i);
      if (c > 127 || c == '"' || c == '<' || c == '>' || c == '&') {
        out.append("&#");
        out.append((int) c);
        out.append(';');
      } else {
        out.append(c);
      }
    }
    return out.toString();
  }

  @SuppressFBWarnings
  public static String padRight(final String s, final int n) {
    return String.format("%1$-" + n + "s", s);
  }
}
