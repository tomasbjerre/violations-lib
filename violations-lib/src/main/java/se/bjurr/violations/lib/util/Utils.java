package se.bjurr.violations.lib.util;

import static se.bjurr.violations.lib.util.StringUtils.padRight;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.DirectoryStream;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Scanner;
import java.util.Set;
import java.util.stream.Collectors;
import se.bjurr.violations.lib.model.Violation;
import se.bjurr.violations.lib.reports.Parser;
import se.bjurr.violations.lib.reports.Reporter;

public class Utils {
  public static <T> T checkNotNull(final T reference, final String errorMessage) {
    if (reference == null) {
      throw new NullPointerException(errorMessage);
    }
    return reference;
  }

  public static String emptyToNull(final String str) {
    if (str == null || str.isEmpty()) {
      return null;
    }
    return str;
  }

  public static String nullToEmpty(final String nullableReference) {
    return nullableReference == null ? "" : nullableReference;
  }

  public static <T> T firstNonNull(final T f, final T s) {
    if (f != null) {
      return f;
    }
    return s;
  }

  @SuppressWarnings("static-access")
  public static InputStream getResource(final String filename) {
    return Thread.currentThread().getContextClassLoader().getSystemResourceAsStream(filename);
  }

  public static boolean isNullOrEmpty(final String str) {
    return str == null || str.isEmpty();
  }

  @SuppressWarnings("resource")
  public static String toString(final InputStream inputStream) throws IOException {
    final Scanner scanner =
        new Scanner(inputStream, StandardCharsets.UTF_8.displayName()).useDelimiter("\\A");
    final String result = scanner.hasNext() ? scanner.next() : "";
    scanner.close();
    inputStream.close();

    return result;
  }

  @SuppressFBWarnings("URLCONNECTION_SSRF_FD")
  public static String toString(final URL resource) throws IOException {
    try {
      return toString(resource.openStream());
    } catch (final IOException e) {
      throw e;
    }
  }

  public static Set<Violation> setReporter(final Set<Violation> violations, final String reporter) {
    for (final Violation v : violations) {
      v.setReporter(reporter);
    }
    return violations;
  }

  public static void updateReadmeWithReporters() throws IOException {

    String reporters = "";
    final List<Reporter> sorted =
        Arrays.stream(Reporter.values())
            .sorted(Comparator.comparing((r1) -> r1.getName()))
            .collect(Collectors.toList());
    for (final Reporter reporter : sorted) {
      reporters +=
          "| "
              + padRight("[_" + reporter.getName() + "_](" + reporter.getUrl() + ") ", 85)
              + " | "
              + padRight("`" + reporter.getParser().name() + "`", 20)
              + " | "
              + reporter.getNote().replace("|", "\\|")
              + "\n";
    }

    final Path readmeFile = findReadmeFile(FileSystems.getDefault().getPath("."));
    if (readmeFile == null) {
      return;
    }
    final String content = new String(Files.readAllBytes(readmeFile), StandardCharsets.UTF_8);
    final String beginPart = "| Reporter | Parser | Notes";
    final String endPart =
        "Missing a format? Open an issue [here](https://github.com/tomasbjerre/violations-lib/issues)!";
    final int start = content.indexOf(beginPart);
    if (start == -1) {
      throw new RuntimeException(
          "Could not find:\n\n" + beginPart + "\n" + endPart + "\n\n in " + readmeFile.toString());
    }
    final int end = content.indexOf(endPart);
    final String beforePart = content.substring(0, start + beginPart.length());
    final String afterPart = content.substring(end);
    final String reportersPart = reporters.trim();
    final String stats =
        Parser.values().length + " parsers and " + Reporter.values().length + " reporters.";
    final String newContent =
        beforePart + "\n| --- | --- | ---\n" + reportersPart + "\n\n" + stats + "\n\n" + afterPart;

    Files.write(readmeFile, newContent.getBytes(StandardCharsets.UTF_8));
  }

  public static Path findReadmeFile(final Path file) throws IOException {
    if (file == null) {
      return null;
    }
    final Set<Path> files = new HashSet<>();
    try (DirectoryStream<Path> stream = Files.newDirectoryStream(file)) {
      for (final Path path : stream) {
        if (!Files.isDirectory(path)) {
          files.add(path);
        }
      }
    }
    for (final Path candidate : files) {
      final Path fileName = candidate.getFileName();
      if (fileName != null && fileName.toString().equals("README.md")) {
        return candidate;
      }
    }
    return findReadmeFile(file.getParent());
  }
}
