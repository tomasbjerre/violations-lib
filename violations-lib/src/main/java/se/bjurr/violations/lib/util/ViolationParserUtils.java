package se.bjurr.violations.lib.util;

import static java.lang.Integer.parseInt;
import static java.util.Optional.empty;
import static java.util.Optional.ofNullable;
import static java.util.regex.Pattern.quote;

import java.io.InputStream;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import javax.xml.XMLConstants;
import javax.xml.stream.XMLInputFactory;
import javax.xml.stream.XMLStreamException;
import javax.xml.stream.XMLStreamReader;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerConfigurationException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.TransformerFactoryConfigurationError;
import javax.xml.transform.stax.StAXSource;
import javax.xml.transform.stream.StreamResult;

public final class ViolationParserUtils {

  private ViolationParserUtils() {}

  public static String asString(final XMLStreamReader xmlr) throws Exception {
    final Transformer transformer = createTranformer();
    final StringWriter stringWriter = new StringWriter();
    transformer.transform(new StAXSource(xmlr), new StreamResult(stringWriter));
    return stringWriter.toString();
  }

  public static Optional<String> findAttribute(final XMLStreamReader in, final String attribute) {
    return ofNullable(in.getAttributeValue("", attribute));
  }

  public static Optional<Integer> findIntegerAttribute(
      final XMLStreamReader in, final String attribute) {
    final String attr = in.getAttributeValue("", attribute);
    if (attr == null) {
      return empty();
    } else {
      return ofNullable(Integer.parseInt(attr));
    }
  }

  public static String getAttribute(final XMLStreamReader in, final String attribute) {
    final String foundOpt = in.getAttributeValue("", attribute);
    if (foundOpt == null) {
      String foundin;
      try {
        foundin = asString(in);
      } catch (final Exception e) {
        throw new RuntimeException(e);
      }
      throw new RuntimeException("\"" + attribute + "\" not found in:\n" + foundin);
    }
    return foundOpt;
  }

  public static Integer getIntegerAttribute(final XMLStreamReader in, final String attribute) {
    return parseInt(getAttribute(in, attribute));
  }

  public static List<String> getLines(final String string) {
    return Arrays.asList(string.split("\n"));
  }

  /**
   * @return List per line in String, with groups from regexpPerLine.
   */
  public static List<List<String>> getLines(final String string, final String regexpPerLine) {
    final List<List<String>> results = new ArrayList<>();
    final Pattern pattern = Pattern.compile(regexpPerLine);
    for (final String line : string.split("\n")) {
      final List<String> found = getLineParts(pattern, line);
      if (!found.isEmpty()) {
        results.add(found);
      }
    }
    return results;
  }

  private static List<String> getLineParts(final Pattern pattern, final String line) {
    final Matcher matcher = pattern.matcher(line);
    if (!matcher.find()) {
      return new ArrayList<>();
    }
    final List<String> lineParts = new ArrayList<>();
    for (int g = 0; g <= matcher.groupCount(); g++) {
      lineParts.add(matcher.group(g));
    }
    return lineParts;
  }

  /**
   * Match one regexp at a time. Remove the matched part from the string, trim, and match next
   * regexp on that string...
   */
  public static List<String> getParts(String string, final String... regexpList) {
    final List<String> parts = new ArrayList<>();
    for (final String regexp : regexpList) {
      final Pattern pattern = Pattern.compile(regexp);
      final Matcher matcher = pattern.matcher(string);
      final boolean found = matcher.find();
      if (!found) {
        return new ArrayList<>();
      }
      final String part = matcher.group(1).trim();
      parts.add(part);
      string = string.replaceFirst(quote(matcher.group()), "").trim();
    }
    return parts;
  }

  private static Transformer createTranformer()
      throws TransformerFactoryConfigurationError, TransformerConfigurationException {
    final TransformerFactory factory = TransformerFactory.newInstance();
    factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
    factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_STYLESHEET, "");
    final Transformer transformer = factory.newTransformer();
    transformer.setOutputProperty(OutputKeys.INDENT, "yes");
    return transformer;
  }

  public static XMLStreamReader createXmlReader(final InputStream input) throws XMLStreamException {
    final XMLInputFactory factory = XMLInputFactory.newInstance();
    factory.setProperty(XMLInputFactory.SUPPORT_DTD, false);
    return factory.createXMLStreamReader(input);
  }
}
