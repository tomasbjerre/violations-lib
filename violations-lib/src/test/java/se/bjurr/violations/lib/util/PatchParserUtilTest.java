package se.bjurr.violations.lib.util;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.Map;
import java.util.Optional;
import java.util.logging.Logger;
import org.approvaltests.Approvals;
import org.junit.Test;

public class PatchParserUtilTest {
  private static Logger LOG = Logger.getLogger(PatchParserUtil.class.getSimpleName());

  private static final String NEW_DIFF =
      "@@ -1,6 +1,6 @@\n <html>\n  <head></head>\n <body>\n-<font>\n+<font> \n </body> \n </html>";
  private static final String CHANGED_DIFF =
      " @@ -1,4 +1,5 @@\n .klass {\n  font-size: 14px;\n+ \n  font-size: 14px;\n }";
  private static final String CHANGED_DIFF_2 =
      "@@ -6,6 +6,16 @@\n"
          + "  void npe(String a, String b) {\n"
          + "   if (a == null) {\n"
          + "    System.out.println();\n"
          + "+   System.out.println();\n"
          + "+  } else {\n"
          + "+\n"
          + "+  }\n"
          + "+  a.length();\n"
          + "+ }\n"
          + "+\n"
          + "+ void npe2(String a, String b) {\n"
          + "+  if (a == null) {\n"
          + "+   System.out.println();\n"
          + "   } else {\n"
          + " \n"
          + "   }\n"
          + "@@ -14,6 +24,6 @@ void npe(String a, String b) {\n"
          + " \n"
          + "  @Override\n"
          + "  public boolean equals(Object obj) {\n"
          + "-  return true;\n"
          + "+  return false;\n"
          + "  }\n"
          + " }";

  @Test
  public void testThatChangedContentCanBeCommented() {
    assertThat(this.findLineToComment("patch", 1)) //
        .isNull();
  }

  @Test
  public void testThatChangedContentCanBeCommentedNewFile() {
    assertThat(this.findLineToComment(NEW_DIFF, 1)) //
        .isEqualTo(1);

    assertThat(this.findLineToComment(NEW_DIFF, 5)) //
        .isEqualTo(6);
  }

  @Test
  public void testThatChangedContentCanBeCommentedChangedFile() {
    assertThat(this.findLineToComment(CHANGED_DIFF, 1)) //
        .isEqualTo(2);

    assertThat(this.findLineToComment(CHANGED_DIFF, 4)) //
        .isEqualTo(5);
  }

  @Test
  public void testThatChangedContentCanBeCommentedChangedPartsOfFile() {
    assertThat(this.findLineToComment(CHANGED_DIFF_2, 6)) //
        .isEqualTo(1);

    assertThat(this.findLineToComment(CHANGED_DIFF_2, 8)) //
        .isEqualTo(3);

    assertThat(this.findLineToComment(CHANGED_DIFF_2, 14)) //
        .isEqualTo(9);

    assertThat(this.findLineToComment(CHANGED_DIFF_2, 21)) //
        .isEqualTo(16);
  }

  @Test
  public void testThatViolationBetweenDiffBlocks() {
    String patch =
        "@@ -143,7 +144,6 @@ import ru.novikov.somepackage1\n"
            + " import ru.novikov.somepackage2\n"
            + " import ru.novikov.somepackage3\n"
            + " import ru.novikov.somepackage4\n"
            + "-import ru.novikov.somepackage5\n"
            + " import ru.novikov.somepackage6\n"
            + " import ru.novikov.somepackage7\n"
            + " import ru.novikov.somepackage8\n"
            + "@@ -187,7 +187,8 @@ import javax.inject.Singleton\n"
            + "             SomeModule1::class,\n"
            + "             SomeModule2::class,\n"
            + "             SomeModule3::class,\n"
            + "-            SomeModule4::class\n"
            + "+            SomeModule4::class,\n"
            + "+            LoggerModule::class\n"
            + "         ]\n"
            + " )\n"
            + " @Singleton\n";

    final PatchParserUtil pp = new PatchParserUtil(patch);
    assertThat(pp.isLineInDiff(150)) //
        .isFalse();
  }

  @Test
  public void testThatOldLineIsEmptyIfOutsideOfDiff() {
    final String patch =
        "--- a/src/main/java/se/bjurr/violations/lib/example/OtherClass.java\n+++ b/src/main/java/se/bjurr/violations/lib/example/OtherClass.java\n@@ -4,12 +4,15 @@ package se.bjurr.violations.lib.example;\n  * No ending dot\n  */\n public class OtherClass {\n- public static String CoNstANT = \"yes\";\n+ public static String CoNstANT = \"yes\"; \n \n  public void myMethod() {\n   if (CoNstANT.equals(\"abc\")) {\n \n   }\n+  if (CoNstANT.equals(\"abc\")) {\n+\n+  }\n  }\n \n  @Override\n";
    Approvals.verify(new PatchParserUtil(patch));

    this.getIntegerOptionalMap(patch);

    final PatchParserUtil pp = new PatchParserUtil(patch);

    assertThat(pp.isLineInDiff(999)) //
        .isFalse();
    assertThat(pp.findOldLine(999).orElse(null)) //
        .isNull();
    assertThat(pp.findLineInDiff(999).orElse(null)) //
        .isNull();

    assertThat(pp.isLineInDiff(6)) //
        .isTrue();
    assertThat(pp.findOldLine(6).orElse(null)) //
        .isEqualTo(6);
    assertThat(pp.findLineInDiff(6).orElse(null)) //
        .isEqualTo(5);
  }

  @Test
  public void testThatLineTableCanBeRetrieved() {
    final String patch =
        "--- a/src/main/java/se/bjurr/violations/lib/example/OtherClass.java\n+++ b/src/main/java/se/bjurr/violations/lib/example/OtherClass.java\n@@ -4,12 +4,15 @@ package se.bjurr.violations.lib.example;\n  * No ending dot\n  */\n public class OtherClass {\n- public static String CoNstANT = \"yes\";\n+ public static String CoNstANT = \"yes\"; \n \n  public void myMethod() {\n   if (CoNstANT.equals(\"abc\")) {\n \n   }\n+  if (CoNstANT.equals(\"abc\")) {\n+\n+  }\n  }\n \n  @Override\n";
    Approvals.verify(new PatchParserUtil(patch));
    final Map<Integer, Optional<Integer>> map = this.getIntegerOptionalMap(patch);

    assertThat(map.get(6).orElse(null)) //
        .isEqualTo(6);
    assertThat(map.get(7).orElse(null)) //
        .isNull();
    assertThat(map.get(8).orElse(null)) //
        .isEqualTo(8);

    assertThat(map.get(12).orElse(null)) //
        .isEqualTo(12);
    assertThat(map.get(13).orElse(null)) //
        .isNull();
    assertThat(map.get(14).orElse(null)) //
        .isNull();
    assertThat(map.get(15).orElse(null)) //
        .isNull();
    assertThat(map.get(16).orElse(null)) //
        .isEqualTo(13);
  }

  @Test
  public void testThatLineTableCanBeRetrieved2() {
    final String patch =
        "--- a/src/main/java/se/bjurr/violations/lib/example/MyClass.java\n+++ b/src/main/java/se/bjurr/violations/lib/example/MyClass.java\n@@ -9,6 +9,8 @@ public class MyClass {\n   } else {\n \n   }\n+  if (a == null)\n+   a.charAt(123);\n   a.length();\n  }\n \n";
    Approvals.verify(new PatchParserUtil(patch));

    final Map<Integer, Optional<Integer>> map = this.getIntegerOptionalMap(patch);

    assertThat(map.get(11).orElse(null)) //
        .isEqualTo(11);
    assertThat(map.get(12).orElse(null)) //
        .isNull();
    assertThat(map.get(13).orElse(null)) //
        .isNull();
    assertThat(map.get(14).orElse(null)) //
        .isEqualTo(12);
  }

  @Test
  public void testPatchApprovalNewDiff() {
    Approvals.verify(new PatchParserUtil(NEW_DIFF));
  }

  @Test
  public void testPatchApprovalChangedDiff() {
    Approvals.verify(new PatchParserUtil(CHANGED_DIFF));
  }

  @Test
  public void testPatchApprovalChangedDiff2() {
    Approvals.verify(new PatchParserUtil(CHANGED_DIFF_2));
  }

  private Integer findLineToComment(final String patch, final int commentLint) {
    this.getIntegerOptionalMap(patch);

    return new PatchParserUtil(patch) //
        .findLineInDiff(commentLint) //
        .orElse(null);
  }

  private Map<Integer, Optional<Integer>> getIntegerOptionalMap(final String patch) {
    final String[] diffLines = patch.split("\n");
    final StringBuilder sb = new StringBuilder();
    for (int i = 0; i < diffLines.length; i++) {
      sb.append(i + 1 + " | " + diffLines[i] + "\n");
    }
    final Map<Integer, Optional<Integer>> map =
        new PatchParserUtil(patch) //
            .getNewLineToOldLineTable();
    for (final Map.Entry<Integer, Optional<Integer>> e : map.entrySet()) {
      sb.append(e.getKey() + " : " + e.getValue().orElse(null) + "\n");
    }
    LOG.info("\n" + sb.toString());
    return map;
  }
}
