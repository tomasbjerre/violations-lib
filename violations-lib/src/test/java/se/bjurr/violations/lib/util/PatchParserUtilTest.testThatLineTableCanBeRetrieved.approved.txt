patch:
0	: --- a/src/main/java/se/bjurr/violations/lib/example/OtherClass.java
1	: +++ b/src/main/java/se/bjurr/violations/lib/example/OtherClass.java
2	: @@ -4,12 +4,15 @@ package se.bjurr.violations.lib.example;
3	:   * No ending dot
4	:   */
5	:  public class OtherClass {
6	: - public static String CoNstANT = "yes";
7	: + public static String CoNstANT = "yes"; 
8	:  
9	:   public void myMethod() {
10	:    if (CoNstANT.equals("abc")) {
11	:  
12	:    }
13	: +  if (CoNstANT.equals("abc")) {
14	: +
15	: +  }
16	:   }
17	:  
18	:   @Override


newLineToLineInDiffTable:
-1 : 2
4 : 3
5 : 4
6 : 5
7 : 7
8 : 8
9 : 9
10 : 10
11 : 11
12 : 12
13 : 13
14 : 14
15 : 15
16 : 16
17 : 17
18 : 18
19 : 19


newLineToOldLineTable:
4 : 4
5 : 5
6 : 6
7 : null
8 : 8
9 : 9
10 : 10
11 : 11
12 : 12
13 : null
14 : null
15 : null
16 : 13
17 : 14
18 : 15
